// @decorators/express/lib/src/middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorMiddlewareHandler = exports.ERROR_MIDDLEWARE = exports.middlewareHandler = void 0;
const di_1 = require("@decorators/di");
/**
 * Create request middleware handler that uses class or function provided as middleware
 */
function middlewareHandler(middleware) {
    return (req, res, next) => {
        invokeMiddleware(middleware, [req, res, next]).catch(next);
    };
}
exports.middlewareHandler = middlewareHandler;
/**
 * Error Middleware class registration DI token
 */
exports.ERROR_MIDDLEWARE = new di_1.InjectionToken('ERROR_MIDDLEWARE');
/**
 * Add error middleware to the app
 */
function errorMiddlewareHandler() {
    return (error, req, res, next) => {
        invokeMiddleware(exports.ERROR_MIDDLEWARE, [error, req, res, next]).catch(next);
    };
}
exports.errorMiddlewareHandler = errorMiddlewareHandler;
/**
 * Instantiate middleware and invoke it with arguments
 */
async function invokeMiddleware(middleware, args) {
    var _a, _b;
    const next = args[args.length - 1];
    try {
        let instance;
        if (typeof middleware === 'function') {
            if ((_a = middleware.prototype) === null || _a === void 0 ? void 0 : _a.use) {
                instance = new middleware(...args);
            }
            else {
                instance = middleware;
            }
        }
        else {
            instance = await di_1.Container.get(middleware);
        }
        const handler = (_b = instance === null || instance === void 0 ? void 0 : instance.use) !== null && _b !== void 0 ? _b : instance;
        const result = typeof handler === 'function' ? handler.apply(instance, args) : instance;
        if (result instanceof Promise) {
            result.catch(next);
        }
    }
    catch (err) {
        next(err);
    }
}
